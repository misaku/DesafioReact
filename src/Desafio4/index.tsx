import React, {useEffect, useState} from 'react';

/* TODO
  Existe neste componente alguns problemas.

  1. A listagem não para de atualizar descontroladamente e exponencialmente.
  2. O total só mostra o valor real quando atualizado os estados das contagens 2 e 3 mas precisa considerar
     qualquer uma das 3 contagens.

  Critérios de aceite:

  - Não usar mais de um useEffect.
  - O cálculo do total deverá ser feito no useEffect.
  - As alterações do estados da lista e total deverá ocorrer no useEffect.
*/

export const Desafio4 = () => {
  const [count1, setCount1] = useState(0)
  const [count2, setCount2] = useState(0)
  const [count3, setCount3] = useState(0)
  const [list, setList] = useState<number[]>([])
  const [total, setTotal] = useState(0)

  useEffect(()=>{
    const totalCount = count1 + count2 + count3;
    setTotal(totalCount)
    setList([...list, totalCount])
  },[count3, count2, list])

  const addCount1 = ()=>setCount1(count1 + 1)
  const addCount2 = ()=>setCount2(count2 + 1)
  const addCount3 = ()=>setCount3(count3 + 1)

  return (
    <>
      <button onClick={addCount1} data-testid="btn1">adiciona + 1 na contagem 1</button><br/>
      <button onClick={addCount2} data-testid="btn2">adiciona + 1 na contagem 2</button><br/>
      <button onClick={addCount3} data-testid="btn3">adiciona + 1 na contagem 3</button><br/>
      <p data-testid="v1">v1: {count1}</p>
      <p data-testid="v2">v2: {count2}</p>
      <p data-testid="v3">v3: {count3}</p>
      <p data-testid="total">total: {total}</p>
      <p data-testid="length">tamanho da lista: {list.length}</p>
    </>
  )
}
