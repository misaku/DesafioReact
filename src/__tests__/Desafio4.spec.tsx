import {Desafio4} from '../Desafio4';
import {waitFor, render, getNodeText} from '@testing-library/react'
import user from '@testing-library/user-event'
describe('Desafio4',()=>{
  it('Deve ser Renderizado com os valores corretos 1',async ()=>{
    const  component = render(<Desafio4 />)
    const btn1 =await component.getByTestId('btn1')
    const btn2 =await component.getByTestId('btn2')
    const btn3 =await component.getByTestId('btn3')
    const length = await component.getByTestId('length')
    const v1 = await component.getByTestId('v1')
    const v2 = await component.getByTestId('v2')
    const v3 = await component.getByTestId('v3')
    const total = await component.getByTestId('total')
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn1)
    await user.click(btn2)
    await user.click(btn2)
    await user.click(btn2)
    await user.click(btn3)
    await user.click(btn3)
    await user.click(btn1)
    await user.click(btn1)
    await waitFor(()=>expect(getNodeText(length)).toEqual('tamanho da lista: 11'))
    expect(getNodeText(v1)).toEqual('v1: 5')
    expect(getNodeText(v2)).toEqual('v2: 3')
    expect(getNodeText(v3)).toEqual('v3: 2')
    expect(getNodeText(total)).toEqual('total: 10')
    expect(component).toMatchSnapshot()
  })
  it('Deve ser Renderizado com os valores corretos 2',async ()=>{
    const  component = render(<Desafio4 />)
    const btn1 =await component.getByTestId('btn1')
    const btn2 =await component.getByTestId('btn2')
    const btn3 =await component.getByTestId('btn3')
    const length = await component.getByTestId('length')
    const v1 = await component.getByTestId('v1')
    const v2 = await component.getByTestId('v2')
    const v3 = await component.getByTestId('v3')
    const total = await component.getByTestId('total')
    await user.click(btn1)
    await user.click(btn2)
    await user.click(btn2)
    await user.click(btn3)
    await user.click(btn3)
    await user.click(btn1)
    await user.click(btn1)
    await waitFor(()=>expect(getNodeText(length)).toEqual('tamanho da lista: 8'))
    expect(getNodeText(v1)).toEqual('v1: 3')
    expect(getNodeText(v2)).toEqual('v2: 2')
    expect(getNodeText(v3)).toEqual('v3: 2')
    expect(getNodeText(total)).toEqual('total: 7')
    expect(component).toMatchSnapshot()
  })
})
