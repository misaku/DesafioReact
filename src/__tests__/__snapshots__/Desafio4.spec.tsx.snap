// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desafio4 Deve ser Renderizado com os valores corretos 1 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        data-testid="btn1"
      >
        adiciona + 1 na contagem 1
      </button>
      <br />
      <button
        data-testid="btn2"
      >
        adiciona + 1 na contagem 2
      </button>
      <br />
      <button
        data-testid="btn3"
      >
        adiciona + 1 na contagem 3
      </button>
      <br />
      <p
        data-testid="v1"
      >
        v1: 
        5
      </p>
      <p
        data-testid="v2"
      >
        v2: 
        3
      </p>
      <p
        data-testid="v3"
      >
        v3: 
        2
      </p>
      <p
        data-testid="total"
      >
        total: 
        10
      </p>
      <p
        data-testid="length"
      >
        tamanho da lista: 
        11
      </p>
    </div>
  </body>,
  "container": <div>
    <button
      data-testid="btn1"
    >
      adiciona + 1 na contagem 1
    </button>
    <br />
    <button
      data-testid="btn2"
    >
      adiciona + 1 na contagem 2
    </button>
    <br />
    <button
      data-testid="btn3"
    >
      adiciona + 1 na contagem 3
    </button>
    <br />
    <p
      data-testid="v1"
    >
      v1: 
      5
    </p>
    <p
      data-testid="v2"
    >
      v2: 
      3
    </p>
    <p
      data-testid="v3"
    >
      v3: 
      2
    </p>
    <p
      data-testid="total"
    >
      total: 
      10
    </p>
    <p
      data-testid="length"
    >
      tamanho da lista: 
      11
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Desafio4 Deve ser Renderizado com os valores corretos 2 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        data-testid="btn1"
      >
        adiciona + 1 na contagem 1
      </button>
      <br />
      <button
        data-testid="btn2"
      >
        adiciona + 1 na contagem 2
      </button>
      <br />
      <button
        data-testid="btn3"
      >
        adiciona + 1 na contagem 3
      </button>
      <br />
      <p
        data-testid="v1"
      >
        v1: 
        3
      </p>
      <p
        data-testid="v2"
      >
        v2: 
        2
      </p>
      <p
        data-testid="v3"
      >
        v3: 
        2
      </p>
      <p
        data-testid="total"
      >
        total: 
        7
      </p>
      <p
        data-testid="length"
      >
        tamanho da lista: 
        8
      </p>
    </div>
  </body>,
  "container": <div>
    <button
      data-testid="btn1"
    >
      adiciona + 1 na contagem 1
    </button>
    <br />
    <button
      data-testid="btn2"
    >
      adiciona + 1 na contagem 2
    </button>
    <br />
    <button
      data-testid="btn3"
    >
      adiciona + 1 na contagem 3
    </button>
    <br />
    <p
      data-testid="v1"
    >
      v1: 
      3
    </p>
    <p
      data-testid="v2"
    >
      v2: 
      2
    </p>
    <p
      data-testid="v3"
    >
      v3: 
      2
    </p>
    <p
      data-testid="total"
    >
      total: 
      7
    </p>
    <p
      data-testid="length"
    >
      tamanho da lista: 
      8
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
